browser/ui/webui/brave_new_tab/new_tab_page_ui.cc

```diff
@@ -0,0 +1,119 @@
+#include "brave/browser/ntp_background/custom_background_file_manager.h"
+#include "brave/browser/ntp_background/view_counter_service_factory.h"
+#include "brave/browser/ui/webui/brave_new_tab/custom_image_chooser.h"
+#include "brave/browser/ui/webui/brave_new_tab/new_tab_page_handler.h"
+#include "brave/browser/ui/webui/brave_webui_source.h"
+#include "brave/components/brave_new_tab/new_tab_prefs.h"
+#include "brave/components/brave_new_tab/resources/grit/brave_new_tab_generated_map.h"
+#include "brave/components/l10n/common/localization_util.h"
+#include "brave/components/ntp_background_images/browser/ntp_custom_images_source.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/webui/webui_util.h"
+#include "components/grit/brave_components_resources.h"
+#include "components/grit/brave_components_strings.h"
+#include "components/prefs/pref_service.h"
+#include "components/strings/grit/components_strings.h"
+#include "content/public/browser/web_ui.h"
+#include "content/public/browser/web_ui_data_source.h"
+#include "ui/base/webui/web_ui_util.h"
+
+namespace brave_new_tab {
+
+namespace {
+
+static constexpr webui::LocalizedString kStrings[] = {
+    {"backgroundSettingsTitle", IDS_NEW_TAB_BACKGROUND_SETTINGS_TITLE},
+    {"braveBackgroundLabel", IDS_NEW_TAB_BRAVE_BACKGROUND_LABEL},
+    {"customBackgroundLabel", IDS_NEW_TAB_CUSTOM_BACKGROUND_LABEL},
+    {"customBackgroundTitle", IDS_NEW_TAB_CUSTOM_BACKGROUND_LABEL},
+    {"gradientBackgroundLabel", IDS_NEW_TAB_GRADIENT_BACKGROUND_LABEL},
+    {"gradientBackgroundTitle", IDS_NEW_TAB_GRADIENT_BACKGROUND_LABEL},
+    {"photoCreditsText", IDS_NEW_TAB_PHOTO_CREDITS_TEXT},
+    {"randomizeBackgroundLabel", IDS_NEW_TAB_RANDOMIZE_BACKGROUND_LABEL},
+    {"settingsTitle", IDS_NEW_TAB_SETTINGS_TITLE},
+    {"showBackgroundsLabel", IDS_NEW_TAB_SHOW_BACKGROUNDS_LABEL},
+    {"showSponsoredImagesLabel", IDS_NEW_TAB_SHOW_SPONSORED_IMAGES_LABEL},
+    {"solidBackgroundLabel", IDS_NEW_TAB_SOLID_BACKGROUND_LABEL},
+    {"solidBackgroundTitle", IDS_NEW_TAB_SOLID_BACKGROUND_LABEL},
+    {"uploadBackgroundLabel", IDS_NEW_TAB_UPLOAD_BACKGROUND_LABEL},
+};
+
+// Adds support for displaying images stored in the custom background image
+// folder.
+void AddCustomImageDataSource(Profile* profile) {
+  auto* custom_background_service =
+      BraveNTPCustomBackgroundServiceFactory::GetForContext(profile);
+  if (!custom_background_service) {
+    return;
+  }
+  auto source = std::make_unique<ntp_background_images::NTPCustomImagesSource>(
+      custom_background_service);
+  content::URLDataSource::Add(profile, std::move(source));
+}
+
+bool ShouldShowBlankPage(Profile* profile) {
+  if (!profile->IsRegularProfile()) {
+    return false;
+  }
+  auto shows_option = prefs::GetNewTabShowsOption(profile->GetPrefs());
+  return shows_option == prefs::NewTabShowsOption::kBlankpage;
+}
+
+}  // namespace
+
+NewTabPageUI::NewTabPageUI(content::WebUI* web_ui)
```

Comment:
Maybe you haven't got back to this yet, but as discussed this should be `new_tab_page/brave_new_tab_*` with no namespace. The existing file/directory is already correct (for some reason I thought it was different) so the flag should be internal to it as is normally the case for this kind of thing.

Reply:
Follow up issue for folders/namespacing: https://github.com/brave/brave-browser/issues/42563